### DISCUSSION TITLE
USB Permissions

### DISCUSSION THREAD

**Discussion by peteyus** [mvp/uhubctl#337] [Q&A] [answered]
I'm trying to grant USB permissions so I can use `uhubctl` to toggle power on my USB ports. Running with `sudo` it works great, but I'm hitting a wall trying to grant permissions so I can do it without `sudo`. I have tried both rebooting and the `udevadm` command to update permissions and still get the permissions error. Am I missing something obvious in granting permissions? Raspberry Pi 2 B v1.1 if that matters.

```
pi@lego-pi:~ $ uhubctl
No compatible smart hubs detected!
Run with -h to get usage info.
There were permission problems while accessing USB.
To fix this, run this tool as root using 'sudo uhubctl',
or add one or more udev rules like below
to file '/etc/udev/rules.d/52-usb.rules':
SUBSYSTEM=="usb", ATTR{idVendor}=="2001", MODE="0666"
then run 'sudo udevadm trigger --attr-match=subsystem=usb'
pi@lego-pi:~ $ sudo uhubctl
Current status for hub 1-1 [0424:9514, USB 2.00, 5 ports]
  Port 1: 0503 power highspeed enable connect [0424:ec00]
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
  Port 5: 0100 power
pi@lego-pi:~ $ cat /etc/udev/rules.d/52-usb.rules
SUBSYSTEM=="usb", ATTR{idVendor}=="0424", MODE="0666"
```

**Comment by peteyus** [comment]
Additional info it helps, lsusb output:

```
pi@lego-pi:~ $ lsusb
Bus 001 Device 003: ID 0424:ec00 Standard Microsystems Corp. SMSC9512/9514 Fast Ethernet Adapter
Bus 001 Device 002: ID 0424:9514 Standard Microsystems Corp. SMC9514 Hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```

**Comment by peteyus** [comment, marked as answer]
Answered myself, I noticed the last root hub entry from lsusb had a different device ID. Adding this line to my 52-usb.rules file fixed it:

```
SUBSYSTEM=="usb", ATTR{idVendor}=="1d6b", MODE="0666"
```

**Reply by mvp** [reply]
You are using old version of uhubctl. If you used uhubctl 2.4.0, this warning would not be shown (and it is actually irrelevant for you controlling hub 1-1 [0424:9514]).
More here: https://github.com/mvp/uhubctl/releases/tag/v2.4.0 and https://github.com/mvp/uhubctl/commit/9b5efa0ec07cd918b3ca7a03ea63cc817aabcda6.

**Comment by snyper0971** [comment]
I am having an issue getting the Permissions to work on a Raspberry PI 2B running the latest buster. I have 2.4.0 installed. Am I missing something?

```
pi@octopi:~ $ cat /etc/udev/rules.d/52-usb.rules
SUBSYSTEM=="usb", ATTR{idVendor}=="0424", MODE="0666
```
```
pi@octopi:~ $ uhubctl -n 0424 -a on -p 2
There were permission problems while accessing USB.
Follow https://git.io/JIB2Z for a fix!
No compatible devices detected!
Run with -h to get usage info.
```
```
pi@octopi:~ $ sudo uhubctl -n 0424 -a on -p 2
Current status for hub 1-1 [0424:9514, USB 2.00, 5 ports, ppps]
  Port 2: 0100 power
Sent power on request
New status for hub 1-1 [0424:9514, USB 2.00, 5 ports, ppps]
  Port 2: 0100 power
```

**Reply by mvp** [reply]
Did you reboot your device after creating rules file?

**Reply by snyper0971** [reply]
1st I tried `sudo udevadm trigger --attr-match=subsystem=usb` then I tried a reboot. I still get the results above.

**Reply by mvp** [reply]
This is weird. The only way to get this error is if `get_hub_info()` call fails for hub `0424:9514` when not running as root: https://github.com/mvp/uhubctl/blob/69c444b75df587df25762011d2197a2c4238cd7e/uhubctl.c#L708-L712
I recommend to either add few more debugging `printf()` statements around to understand why it fails, or if you have `gdb` installed on your Pi, run uhubctl under `gdb` using something like this:

    $ gdb uhubctl
    gdb> set args -n 0424 -a on -p 2
    gdb> b get_hub_info
    gdb> c
    gdb> # from here use n(ext), s(ub) or p variable to debug

**Reply by mvp** [reply]
@snyper0971:  are you sure you configured it properly? This does not look right:

```
SUBSYSTEM=="usb", ATTR{idVendor}=="0424", MODE="0666
```

It should be (note extra double quote):
```
SUBSYSTEM=="usb", ATTR{idVendor}=="0424", MODE="0666"
```

**Reply by snyper0971** [reply]
Great catch! Always something small and simple.... That did the trick, thank you!

**Comment by ecnelson** [comment]
How about `sudo chmod u+s /usr/sbin/uhubctl`

I have a python daemon running as user 'pi' that occasionally gets IO Errors from an FTDI gadget. The errors persist even after the daemon restarts. It could be pilot error on my part, or the gadget itself may be on its way out (I've had one fail before). In the meantime, what *does* seem to fix it is removing and reinserting the FTDI gadget. Some of these devices (RaspberryPis feeding environmental data) are remote, so `uhubctl` is perfect! Thank you for inventing it.

That said, `sudo` in a python script is a potential issue, so I just set the setuid bits on uhubctl. What could go wrong, eh?

**Reply by mvp** [reply]
Don't set suid, it could be security issue if uhubctl has a bug. Instead, follow readme on how to set USB permissions properly, it will make it work without suid.
